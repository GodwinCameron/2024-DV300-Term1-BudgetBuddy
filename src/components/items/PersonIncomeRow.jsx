import React from "react";

function PersonIncomeRow(props) {
  const { person, index } = props;

  return (
    <div data-testid="person-income-row" className="card card-row p-2 mb-2 row-income">
        {person && (
          <>
            <span>
              <div className="avatar">{person.icon}</div>
              <p>
                {index}. <span>{person.name}</span>
              </p>
            </span>
            <p className="amount">
              {new Intl.NumberFormat("en-za", {
                style: "currency",
                currency: "ZAR",
              }).format(person.salary)}
            </p>
          </>
        )}
    </div>
  );
}

export default PersonIncomeRow;
