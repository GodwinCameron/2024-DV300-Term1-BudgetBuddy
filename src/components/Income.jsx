import React from "react";
import PersonIncomeRow from "./items/PersonIncomeRow";
import { Button, Form } from "react-bootstrap";
import { dummyIncome, iconOptions } from "../utils";

function Income(props) {
  const { income } = props;

  console.log(props.income);
  const handleIncome = (e) => {
    e.preventDefault();
    const icon = e.target.icon.value;
    const name = e.target.name.value;
    const salary = e.target.income.value;
    // tax section
    const taxedAmount = { percent: 0, extra: 0 };
    if (salary <= 237100) {
      taxedAmount.percent = 0.18;
      taxedAmount.extra = 0;
    } else if (salary <= 370500) {
      taxedAmount.percent = 0.26;
      taxedAmount.extra = 42678;
    } else if (salary <= 512800) {
      taxedAmount.percent = 0.31;
      taxedAmount.extra = 77362;
    } else if (salary <= 673000) {
      taxedAmount.percent = 0.36;
      taxedAmount.extra = 121475;
    } else if (salary <= 857900) {
      taxedAmount.percent = 0.39;
      taxedAmount.extra = 179147;
    } else if (salary <= 1817000) {
      taxedAmount.percent = 0.41;
      taxedAmount.extra = 251258;
    } else {
      taxedAmount.percent = 0.45;
      taxedAmount.extra = 644489;
    }
    const tax = {
      percent: taxedAmount.percent,
      extra: taxedAmount.extra,
      amount: salary * taxedAmount.percent + taxedAmount.extra,
    };
    if (icon === "-" || name === "" || salary === "") {
      alert("Please fill out all fields");
      return;
    }
    props.setIncome([...props.income, { icon, name, salary, tax }]);
    e.target.reset();
  };

  return (
    <div>
      <h3>Household Montly Salaries</h3>

      {/* Form */}
      <form data-testid="form" onSubmit={handleIncome} className="form-row">
        <Form.Select
          name="icon"
          defaultValue="-"
          autoComplete="off"
        >
          <option disabled>-</option>
          {iconOptions.map((icon, index) => (
            <option key={index} value={icon}>
              {icon}
            </option>
          ))}
        </Form.Select>
        <Form.Control
          type="text"
          id="name"
          name="name"
          placeholder="Member Name"
          autoComplete="off"
        />
        <Form.Control
          type="number"
          id="income"
          name="income"
          step="0.01"
          placeholder="0.00"
          autoComplete="off"
        />
        <Button type="submit" className="add-income">
          Add
        </Button>
      </form>

      {/* List */}
      <div className="income-outer hide-scroll">
        {/* {dummyIncome.map((item, index) => (
                <PersonIncomeRow key={index} index={index+1} person={item} />
            ))} */}
        {income &&
          income.map((item, index) => (
            <PersonIncomeRow key={index} index={index + 1} person={item} />
          ))}
      </div>
    </div>
  );
}

export default Income;
