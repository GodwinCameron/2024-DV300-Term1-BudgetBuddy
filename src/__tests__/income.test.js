import { fireEvent, render, screen } from "@testing-library/react";
import Income from "../components/Income";
import PersonIncomeRow from "../components/items/PersonIncomeRow";

describe("Tests for Income component", () => {
  test("test component for default state", () => {
    // ARRANGE

    // ACT
    render(<Income />);
    var salariesTitle = screen.getByText("Household Montly Salaries");

    //ASSERT
    expect(salariesTitle).toBeInTheDocument();
  });
});

describe("Tests for PersonIncomeRow component", () => {
    test("test for PersonIncomeRow deault state", () => {
      render(<PersonIncomeRow />);
      var personIncomeRowContainer = screen.getByTestId("person-income-row");
      expect(personIncomeRowContainer).toBeInTheDocument();
    });

    test("test for PersonIncomeRow", () => {
      var index = 1;
        var person = {
            name: "John",
            salary: 5000,
        };
        const { container } = render(<PersonIncomeRow index={index} person={person} />);
        console.log(container.innerHTML);
  
      var personIncomeName = screen.getByText("John");
  
      expect(personIncomeName).toBeInTheDocument();
    });
});

describe("snapshots", () => {
  test("snapshot of income", () => {
    const { asFragment } = render(<Income />);
    expect(asFragment()).toMatchSnapshot();
  });
});