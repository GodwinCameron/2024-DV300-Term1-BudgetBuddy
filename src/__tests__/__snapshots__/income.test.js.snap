// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots snapshot of income 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Household Montly Salaries
    </h3>
    <form
      class="form-row"
      data-testid="form"
    >
      <select
        autocomplete="off"
        class="form-select"
        name="icon"
      >
        <option
          disabled=""
          selected=""
        >
          -
        </option>
        <option
          value="👩"
        >
          👩
        </option>
        <option
          value="👱‍♂️"
        >
          👱‍♂️
        </option>
        <option
          value="🐶"
        >
          🐶
        </option>
        <option
          value="🐱"
        >
          🐱
        </option>
        <option
          value="🐰"
        >
          🐰
        </option>
        <option
          value="🐻"
        >
          🐻
        </option>
        <option
          value="🦁"
        >
          🦁
        </option>
        <option
          value="🐷"
        >
          🐷
        </option>
        <option
          value="🐮"
        >
          🐮
        </option>
      </select>
      <input
        autocomplete="off"
        class="form-control"
        id="name"
        name="name"
        placeholder="Member Name"
        type="text"
      />
      <input
        autocomplete="off"
        class="form-control"
        id="income"
        name="income"
        placeholder="0.00"
        step="0.01"
        type="number"
      />
      <button
        class="add-income btn btn-primary"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="income-outer hide-scroll"
    />
  </div>
</DocumentFragment>
`;
